redhat-appstudio/service-provider-integration-operator
# feat: Implement token matching for Github

### What does this PR do?
This supersedes #37.

This implements token matching algorithm for github.

SPI acts as an OAuth Application within Github. Each user can give the application a set of permissions (expressed using scopes). This is a one-to-one mapping - i.e. the OAuth application can only have one set of permissions for any given user. These permissions give the application capabilities to perform actions on ALL repositories that the user owns.

Furthermore, the user can be a member of organizations. This membership can have various levels, giving the user a certain role in the organization. Depending on this role, the OAuth application might or might not be able to exercise the permissions given to it by the user on the repositories of the organization.

In this PR the implementation doesn't distinguish between owned repos and repos visible due to organization membership and thus can match tokens that in reality don't possess enough permissions for given organization repository.

Note that this PR changes the `SPIAccessToken` CRD - it moves the `TokenMetadata` from the `spec` to the `status` because the metadata is now fully figured out by the operator using the token. The `TokenMetadata` now also contains the actual scopes of the token and service-provider-specific blob that is meant to store additional data that is used for matching the tokens with the bindings in sp-specific way. For Github, this blob contains the list of repositories that the token enables access to along with the repository role held by the token for each of the repositories.

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-38

### How to test this PR?
1. Deploy and configure SPI with a github OAuth app (see the repository README for details)
1. Create a binding that requests only read on webhooks, similar to `samples/binding-read-webhooks.yaml`, for a repo that you have access to.
1. Approve the binding by visiting its OAuth URL.
1. Create a binding that requests read and write on the same repository similar to `samples/binding-write-webhooks.yaml` for the same repository.
1. Check that the second binding is not `Injected` and has a non-empty OAuth URL.
1. When clicking on that URL, Github will ask you if you're OK to grant your OAuth app additional permissions.
1. Upon approving in Github, the second binding should also be injected.

## Additional Information
The `quay.io/lkrejci/service-provider-integration-operator:svpi-38` image is built from the latest commit in this PR.
